pageserver/src/tenant/timeline.rs

```diff
@@ -4338,24 +4383,28 @@ impl Timeline {
                 // because we have not implemented L0 => L0 compaction.
                 duplicated_layers.insert(l.layer_desc().key());
             } else if LayerMap::is_l0(l.layer_desc()) {
-                return Err(CompactionError::Other(anyhow!("compaction generates a L0 layer file as output, which will cause infinite compaction.")));
+                bail!("compaction generates a L0 layer file as output, which will cause infinite compaction.");
             } else {
                 insert_layers.push(l.clone());
             }
         }
 
-        let remove_layers = {
-            let mut deltas_to_compact = deltas_to_compact;
-            // only remove those inputs which were not outputs
-            deltas_to_compact.retain(|l| !duplicated_layers.contains(&l.layer_desc().key()));
-            deltas_to_compact
-        };
+        // only remove those inputs which were not outputs
+        let remove_layers: Vec<Layer> = layers_to_remove
+            .iter()
+            .filter(|l| !duplicated_layers.contains(&l.layer_desc().key()))
+            .cloned()
+            .collect();
+
+        if !new_images.is_empty() {
+            guard.track_new_image_layers(new_images, &self.metrics);
+        }
 
         // deletion will happen later, the layer file manager calls garbage_collect_on_drop
         guard.finish_compact_l0(&remove_layers, &insert_layers, &self.metrics);
 
         if let Some(remote_client) = self.remote_client.as_ref() {
-            remote_client.schedule_compaction_update(&remove_layers, &new_layers)?;
+            remote_client.schedule_compaction_update(&remove_layers, new_deltas)?;
```

Comment:
So previously this could had been a stopping error, but now it's just an anyhow?

Reply:
The patch only changes the argument `new_layers` to `new_deltas`, it's a rename only really.

Reply:
> The patch only changes the argument new_layers to new_deltas, it's a rename only really.

On this line yes, but this PR introduces an additional hop to calling this `RemoteTimelineClient::schedule_compaction_update`. Assuming we've merged christian's PR to use RemoteTimelineClient error types conversions for `CompactError` we no longer will take those paths because the error will be wrapped as anyhow.

Old: `Timeline::compact_level0_phase1() -> Result<(), CompactError>` called `RTC::schedule_compaction_update()`... Aah, I see, we did not merge christians PR.

Oki, I'll resolve.